From 30e079d0b9045f5f83620eca9d50177daaedf69b Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Mon, 22 Aug 2022 15:24:53 +0300
Subject: [PATCH] [SW-101652] gaudi3: pcie: update sram init done condition

Currently, rom polls for value of sram init done to be 0 instead of 1
thats why it remains stuck in infinite while loop. This patch add changes
to poll for 1 instead of 0 and also add timeout of max 100ms instead of
looping infinitely. Timeout in init done is recorded in ROM status
update.

Change-Id: I2751dd59867e4dfa5d7394b624ba613471e63f7c
---

diff --git a/arch/gaudi3/arc/common.c b/arch/gaudi3/arc/common.c
index bc8faba..accd251 100644
--- a/arch/gaudi3/arc/common.c
+++ b/arch/gaudi3/arc/common.c
@@ -145,13 +145,34 @@
 	regmap_writel(1, mmD0_PCIE_PHY_CR_PARA_SEL_3);
 }
 
+#define PHY_SRAM_INIT_UDELAY			100000 /* 100ms */
 void pcie_wait_for_sram_init_done(void)
 {
-	while (regmap_readl(mmD0_PCIE_PHY_PHYX_SRAM_INIT_DONE_0) ||
-		regmap_readl(mmD0_PCIE_PHY_PHYX_SRAM_INIT_DONE_1) ||
-		regmap_readl(mmD0_PCIE_PHY_PHYX_SRAM_INIT_DONE_2) ||
-		regmap_readl(mmD0_PCIE_PHY_PHYX_SRAM_INIT_DONE_3))
-		;
+	uint32_t sram_init_done0, sram_init_done1, sram_init_done2, sram_init_done3;
+	uint64_t start = arcboot_rom_get_ts_us();
+
+	/*
+	 * As per HW team, internal sram initialization can take several ms.
+	 * timeout should be added in 10-100ms.
+	 */
+	while (1) {
+		sram_init_done0 = regmap_readl(mmD0_PCIE_PHY_PHYX_SRAM_INIT_DONE_0);
+		sram_init_done1 = regmap_readl(mmD0_PCIE_PHY_PHYX_SRAM_INIT_DONE_1);
+		sram_init_done2 = regmap_readl(mmD0_PCIE_PHY_PHYX_SRAM_INIT_DONE_2);
+		sram_init_done3 = regmap_readl(mmD0_PCIE_PHY_PHYX_SRAM_INIT_DONE_3);
+
+		if ((sram_init_done0 == D0_PCIE_PHY_PHYX_SRAM_INIT_DONE_STATUS_M) &&
+			(sram_init_done1 == D0_PCIE_PHY_PHYX_SRAM_INIT_DONE_STATUS_M) &&
+			(sram_init_done2 == D0_PCIE_PHY_PHYX_SRAM_INIT_DONE_STATUS_M) &&
+			(sram_init_done3 == D0_PCIE_PHY_PHYX_SRAM_INIT_DONE_STATUS_M))
+			break;
+
+		if (arcboot_rom_timeout_expired(start, PHY_SRAM_INIT_UDELAY)) {
+			arcboot_rom_status_update(
+				PROGRESS_ARCBOOT_ROM_PCIE_PHY_SRAM_INIT_TIMEDOUT);
+			break;
+		}
+	}
 }
 
 void pcie_wa_snps_case_01038891(void)
diff --git a/arch/gaudi3/arc/progress_bits.h b/arch/gaudi3/arc/progress_bits.h
index c8492b2..aed72a9 100644
--- a/arch/gaudi3/arc/progress_bits.h
+++ b/arch/gaudi3/arc/progress_bits.h
@@ -104,40 +104,41 @@
 	PROGRESS_ARCBOOT_IN_ROM_RECOVERY = 97,
 	PROGRESS_ARCBOOT_ROM_PCIE_LINK_START = 98,
 	PROGRESS_ARCBOOT_ROM_PCIE_LINK_INIT = 99,
-	PROGRESS_ARCBOOT_ROM_PCIE_LINK_LIMIT_GEN1 = 100,
-	PROGRESS_ARCBOOT_ROM_PCIE_LINK_IATU_CONFIG = 101,
-	PROGRESS_ARCBOOT_ROM_PCIE_LINK_FINISH = 102,
-	PROGRESS_ARCBOOT_ROM_D2D_LINK_START = 103,
-	PROGRESS_ARCBOOT_ROM_D2D_LINK_STS_BUSY = 104,
-	PROGRESS_ARCBOOT_ROM_D2D_LINK_INIT_TIMEDOUT = 105,
-	PROGRESS_ARCBOOT_ROM_D2D_LINK_INIT_SUCCESS = 106,
-	PROGRESS_ARCBOOT_ROM_RECOVERY_END = 107,
-	PROGRESS_ARCBOOT_IN_EROM_RECOVERY = 108,
-	PROGRESS_INIT_EROM_BOOT_DESC_DMA_START = 109,
-	PROGRESS_INIT_EROM_BOOT_DESC_CRC_FAIL = 110,
-	PROGRESS_INIT_EROM_BOOT_DESC_UNALIGNED_DEST = 111,
-	PROGRESS_INIT_EROM_BOOT_DESC_INVALID_IMG_SIZE = 112,
-	PROGRESS_INIT_EROM_BOOT_DESC_UNALIGNED_IMG_SIZE = 113,
-	PROGRESS_INIT_EROM_BOOT_DESC_DMA_ERROR = 114,
-	PROGRESS_INIT_EROM_BOOT_DESC_DMA_FINISH = 115,
-	PROGRESS_INIT_EROM_BOOT_DESC_HDR_MAGIC_FAIL = 116,
-	PROGRESS_INIT_EROM_BOOT_DESC_VERIFICATION_FAIL = 117,
-	PROGRESS_INIT_EROM_BOOT_DESC_FAIL = 118,
-	PROGRESS_EROM_BOOT_DESC_VERIFY_START = 119,
-	PROGRESS_EROM_BOOT_DESC_VERIFY_VERSION_FAIL = 120,
-	PROGRESS_EROM_BOOT_DESC_VERIFY_KEY_REVOKED = 121,
-	PROGRESS_EROM_BOOT_DESC_VERIFY_KEY_HASH_FAIL = 122,
-	PROGRESS_EROM_BOOT_DESC_VERIFY_AUTH_FAIL = 123,
-	PROGRESS_EROM_BOOT_DESC_VERIFY_SUCCESS = 124,
-	PROGRESS_ARCBOOT_EROM_IMG0 = 125,
-	PROGRESS_ARCBOOT_EROM_IMG0_DMA_START = 126,
-	PROGRESS_ARCBOOT_EROM_IMG0_DMA_CRC_FAILED = 127,
-	PROGRESS_ARCBOOT_EROM_IMG0_DMA_UNALIGNED_DEST = 128,
-	PROGRESS_ARCBOOT_EROM_IMG0_DMA_INVALID_IMG_SIZE = 129,
-	PROGRESS_ARCBOOT_EROM_IMG0_DMA_UNALIGNED_IMG_SIZE = 130,
-	PROGRESS_ARCBOOT_EROM_IMG0_DMA_ERROR = 131,
-	PROGRESS_ARCBOOT_EROM_IMG0_DMA_FINISH = 132,
-	PROGRESS_ARCBOOT_EROM_IMG0_VERIFICATION_FAIL = 133,
-	PROGRESS_ARCBOOT_EROM_IMG0_START = 134,
-	PROGRESS_ARCBOOT_EROM_IMG0_FINISH = 135,
+	PROGRESS_ARCBOOT_ROM_PCIE_PHY_SRAM_INIT_TIMEDOUT = 100,
+	PROGRESS_ARCBOOT_ROM_PCIE_LINK_LIMIT_GEN1 = 101,
+	PROGRESS_ARCBOOT_ROM_PCIE_LINK_IATU_CONFIG = 102,
+	PROGRESS_ARCBOOT_ROM_PCIE_LINK_FINISH = 103,
+	PROGRESS_ARCBOOT_ROM_D2D_LINK_START = 104,
+	PROGRESS_ARCBOOT_ROM_D2D_LINK_STS_BUSY = 105,
+	PROGRESS_ARCBOOT_ROM_D2D_LINK_INIT_TIMEDOUT = 106,
+	PROGRESS_ARCBOOT_ROM_D2D_LINK_INIT_SUCCESS = 107,
+	PROGRESS_ARCBOOT_ROM_RECOVERY_END = 108,
+	PROGRESS_ARCBOOT_IN_EROM_RECOVERY = 109,
+	PROGRESS_INIT_EROM_BOOT_DESC_DMA_START = 110,
+	PROGRESS_INIT_EROM_BOOT_DESC_CRC_FAIL = 111,
+	PROGRESS_INIT_EROM_BOOT_DESC_UNALIGNED_DEST = 112,
+	PROGRESS_INIT_EROM_BOOT_DESC_INVALID_IMG_SIZE = 113,
+	PROGRESS_INIT_EROM_BOOT_DESC_UNALIGNED_IMG_SIZE = 114,
+	PROGRESS_INIT_EROM_BOOT_DESC_DMA_ERROR = 115,
+	PROGRESS_INIT_EROM_BOOT_DESC_DMA_FINISH = 116,
+	PROGRESS_INIT_EROM_BOOT_DESC_HDR_MAGIC_FAIL = 117,
+	PROGRESS_INIT_EROM_BOOT_DESC_VERIFICATION_FAIL = 118,
+	PROGRESS_INIT_EROM_BOOT_DESC_FAIL = 119,
+	PROGRESS_EROM_BOOT_DESC_VERIFY_START = 120,
+	PROGRESS_EROM_BOOT_DESC_VERIFY_VERSION_FAIL = 121,
+	PROGRESS_EROM_BOOT_DESC_VERIFY_KEY_REVOKED = 122,
+	PROGRESS_EROM_BOOT_DESC_VERIFY_KEY_HASH_FAIL = 123,
+	PROGRESS_EROM_BOOT_DESC_VERIFY_AUTH_FAIL = 124,
+	PROGRESS_EROM_BOOT_DESC_VERIFY_SUCCESS = 125,
+	PROGRESS_ARCBOOT_EROM_IMG0 = 126,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_START = 127,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_CRC_FAILED = 128,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_UNALIGNED_DEST = 129,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_INVALID_IMG_SIZE = 130,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_UNALIGNED_IMG_SIZE = 131,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_ERROR = 132,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_FINISH = 133,
+	PROGRESS_ARCBOOT_EROM_IMG0_VERIFICATION_FAIL = 134,
+	PROGRESS_ARCBOOT_EROM_IMG0_START = 135,
+	PROGRESS_ARCBOOT_EROM_IMG0_FINISH = 136,
 };
